pil-analyzer/src/type_inference.rs

```diff
@@ -0,0 +1,717 @@
+        // - analyze each such group in an environment, where their type schemes
+        //   are instantiated once at the start and not anymore for the symbol lookup.
+
+        // Sort the names such that called names occur first.
+        let names = sort_called_first(
+            definitions
+                .iter()
+                .map(|(n, (_, v))| (n.as_str(), v.as_deref())),
+        );
+
+        self.setup_declared_types(definitions);
+
+        // These are the inferred types for symbols that are declared
+        // as type schemes. They are compared to the declared types
+        // at the end.
+        let mut inferred_types: HashMap<String, Type> = Default::default();
+
+        // Now go through all symbols and derive types for the expressions.
+        // While analyzing a symbol, we ignore its declared type (unless the
+        // symbol is referenced). Unifying the declared type with the inferred
+        // type is done at the end.
+        for name in names {
+            // Ignore builtins (removed from definitions) and definitions without value.
+            let Some((_, Some(value))) = definitions.get_mut(&name) else {
+                continue;
+            };
+
+            let declared_type = self.declared_types[&name].clone();
+            let result = if declared_type.vars.is_empty() {
+                self.process_concrete_symbol(&name, declared_type.ty.clone(), value)
+            } else {
+                self.infer_type_of_expression(value).map(|ty| {
+                    inferred_types.insert(name.to_string(), ty);
+                })
+            };
+            if let Err(e) = result {
+                return Err(format!(
+                    "Error type checking the symbol {name} = {value}:\n{e}",
+                ));
+            }
+        }
+
+        self.check_expressions(expressions)?;
+
+        // From this point on, the substitutions are fixed.
+
+        // Now we check for all symbols that are not declared as a type scheme that they
+        // can resolve to a concrete type.
+        for (name, declared_type) in &self.declared_types {
+            if declared_type.vars.is_empty() {
+                // It is not a type scheme, see if we were able to derive a concrete type.
+                let inferred = self.substitute_to(declared_type.ty.clone());
+                if !inferred.is_concrete_type() {
+                    let inferred_scheme = self.to_type_scheme(inferred);
+                    return Err(format!(
+                        "Could not derive a concrete type for symbol {name}.\nInferred type scheme: {}\n",
+                        format_type_scheme_around_name(
+                            name,
+                            &Some(inferred_scheme),
+                        )
+                    ));
+                }
+            }
+        }
+
+        // We check type schemes last, because only at this point do we know
```

Comment:
maybe this last part could be a `check_type_schemes` method that takes in the `inferred_types`?